staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go

```diff
@@ -101,55 +68,100 @@ func WithAudit(handler http.Handler, requestContextMapper request.RequestContext
 
-		username := "<none>"
-		groups := "<none>"
-		if attribs.GetUser() != nil {
-			username = attribs.GetUser().GetName()
-			if userGroups := attribs.GetUser().GetGroups(); len(userGroups) > 0 {
-				groups = auditStringSlice(userGroups)
+		// set the level
+		level := auditinternal.LevelMetadata
+		if policy != nil && len(policy.Rules) > 0 {
+			// This is just a hack to get through the test without setting a high level by default.
+			// TODO: add the policy evalutation here
+			level = policy.Rules[0].Level
+		}
+		headerLevels := req.Header[AuditLevelHeader]
+		if len(headerLevels) > 0 {
+			switch l := auditinternal.Level(headerLevels[0]); l {
+			case auditinternal.LevelMetadata, auditinternal.LevelNone, auditinternal.LevelResponse, auditinternal.LevelRequest:
+				level = l
+			default:
+				glog.Warningf("Invalid audit level in %v helper: %q", AuditLevelHeader, l)
 			}
 		}
-		asuser := req.Header.Get(authenticationapi.ImpersonateUserHeader)
-		if len(asuser) == 0 {
-			asuser = "<self>"
+
+		ev := &auditinternal.Event{
+			Level:     level,
+			Timestamp: metav1.NewTime(time.Now()),
+			AuditID:   types.UID(id),
 		}
-		asgroups := "<lookup>"
-		requestedGroups := req.Header[authenticationapi.ImpersonateGroupHeader]
-		if len(requestedGroups) > 0 {
-			asgroups = auditStringSlice(requestedGroups)
+		ctx = request.WithAuditEvent(ctx, ev)
+		if err := requestContextMapper.Update(req, ctx); err != nil {
+			glog.Errorf("Failed to attach audit event to the context: %v", err)
+			responsewriters.InternalError(w, req, errors.New("failed to update context"))
+			return
 		}
-		namespace := attribs.GetNamespace()
-		if len(namespace) == 0 {
-			namespace = "<none>"
+
+		attribs, err := GetAuthorizerAttributes(ctx)
+		if err != nil {
+			glog.Errorf("Failed to GetAuthorizerAttributes: %v", err)
+			responsewriters.InternalError(w, req, errors.New("failed to parse request"))
+			return
 		}
-		id := uuid.NewRandom().String()
 
-		line := fmt.Sprintf("%s AUDIT: id=%q ip=%q method=%q user=%q groups=%q as=%q asgroups=%q namespace=%q uri=%q\n",
-			time.Now().Format(time.RFC3339Nano), id, utilnet.GetClientIP(req), req.Method, username, groups, asuser, asgroups, namespace, req.URL)
-		if _, err := fmt.Fprint(out, line); err != nil {
-			glog.Errorf("Unable to write audit log: %s, the error is: %v", line, err)
+		if err := audit.UpdateEventFromRequest(ctx, ev, req, attribs); err != nil {
+			glog.Errorf("Failed to complete audit event from request: %v", err)
+			responsewriters.InternalError(w, req, errors.New("failed to update context"))
+			return
 		}
-		respWriter := decorateResponseWriter(w, out, id)
+
+		// intercept the status code
+		longRunning := false
+		var longRunningSink chan<- *auditinternal.Event
+		if longRunningCheck != nil {
+			ri, _ := request.RequestInfoFrom(ctx)
+			if longRunning = longRunningCheck(req, ri); longRunning {
+				longRunningSink = sink
+			}
+		}
+		respWriter := decorateResponseWriter(w, ev, longRunningSink)
```

Comment:
because we only send the first event for long-running requests. See discussion in comment below. We can change this.